pkg/pma/pma.c

```diff
@@ -0,0 +1,783 @@
+#include "pma.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "page.h"
+#include "sigsegv.h"
+#include "util.h"
+#include "wal.h"
+
+#ifdef PMA_TEST
+#    define inline_
+#    define static_
+#else
+#    define inline_ inline
+#    define static_ static
+#endif
+
+#define PMA_DEBUG
+
+//==============================================================================
+// CONSTANTS
+
+/// Number of bits in a byte.
+static_ const size_t kBitsPerByte = 8;
+
+/// Number of bits required to track the status of a page.
+static_ const size_t kBitsPerPage = 2;
+
+/// Number of pages whose status can be tracked in a single byte.
+static_ const size_t kPagesPerByte = kBitsPerByte / kBitsPerPage;
+
+/// Number of bytes in a single entry of the pg_status array of pma_t.
+static_ const size_t kBytesPerEntry = sizeof(*((pma_t *)NULL)->pg_status);
+
+/// Number of pages whose status can be tracked by a single entry in the
+/// pg_status array of pma_t.
+static_ const size_t kPagesPerEntry = kBytesPerEntry * kPagesPerByte;
+
+/// Extension appended to a backing file to create that file's wal name.
+static_ const char kWriteAheadLogExtension[] = ".wal";
+
+//==============================================================================
+// GLOBAL VARIABLES
+
+/// Global libsigsegv dispatcher.
+static_ sigsegv_dispatcher dispatcher;
+
+//==============================================================================
+// STATIC FUNCTIONS
+
+/// Determine the 0-based page index of an address within the bounds of a PMA
+/// relative to the start of the heap.
+///
+/// @param[in] addr  Address to determine page index for.
+/// @param[in] pma
+static_ inline_ size_t
+addr_to_page_idx_(const void *addr, const pma_t *pma)
```

Comment:
Another noob question: why end these function names with `_`s?

Reply:
The trailing `_` is used for identifiers that are statically scoped. Static functions, static globals, and static types are all suffixed with a trailing `_`.